hasktorch/src/Torch/Random.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Torch.Random where
+
+import ATen.Cast
+import ATen.Class (Castable(..))
+import qualified ATen.Const as ATen
+import qualified ATen.Type as ATen
+import Data.Word
+import qualified ATen.Managed.Type.Generator as ATen
+import qualified Torch.Managed.Native as LibTorch
+import Torch.Device
+import Torch.TensorOptions
+import Torch.Tensor
+import Foreign.ForeignPtr
+import System.IO.Unsafe
+
+data Generator
+  = Generator
+  { device :: Device
+  , seed :: Word64
+  }
+  deriving (Eq, Show)
+
+cpuGenerator :: Word64 -> Generator
+cpuGenerator seed = Generator (Device CPU 0) seed
+
+type RandomGenFunc = ForeignPtr ATen.IntArray -> ForeignPtr ATen.Generator -> ForeignPtr ATen.TensorOptions -> IO (ForeignPtr ATen.Tensor)
+
+generatorFactory :: RandomGenFunc -> [Int] -> TensorOptions -> Generator -> (Tensor,Generator)
+generatorFactory func size options Generator{..} =
+  unsafePerformIO $ do
+    gen <- case device of
+             Device CPU _ -> ATen.newCPUGenerator seed
+             Device CUDA idx -> do
+               gen <- ATen.newCUDAGenerator (fromIntegral idx)
+               ATen.generator_set_current_seed gen seed
+               return gen
+    tensor <- (cast3 func) size gen options
+    nextSeed <- ATen.generator_current_seed gen
+    return (tensor,Generator device nextSeed)
```

Comment:
would it be possible to add the `ATen.Generator` to the `Torch.Random.Generator` data type?

Reply:
I think it it not possible, because `ATen.Generator` is mutable.
In following case, the result depends on whether t2 or t3 is evaluated first, if we add it.

```
let (t,next) = randn [2] (cpuGenerator 0)
     (t2,next2) = randn [2] next
     (t3,next2) = randn [2] next

print t2
print t3
-- or --
print t3
print t2
```

Reply:
I'm worried about allocating new generators all the time when we could reuse already existing ones.
what if you made the constructor for `Torch.Random.Generator` private, and forced people to use a few smart utility functions that always make sure that the internal `ATen.Generator` is set to the accompanying seed in `Torch.Random.Generator`. that way you would prevent the tracked seed and the internal seed to run out of sync. kind of like you are already doing in the `generatorFactory` function...

Reply:
When the random-tensor's size is big enough, we can ignore the overhead.
However, I think  a implementation of the tracked seed by using following data.

```
data Generator = Generator {
  seed :: Word64
, device :: Device
, generator :: (TVar (Maybe (ForeignPtr Generator)))
```

Reply:
As you said, privete constructor is a good choice anyway.
I will add it and hide original constructor.

Reply:
cool, looking forward to that!